### PR TITLE
avm2: Make Matrix3D.*transformVector*() native

### PR DESCRIPTION
**Pull request by kjarosh**
## Description

This makes the following 3 methods native:
* Matrix3D.deltaTransformVector(),
* Matrix3D.transformVector(),
* Matrix3D.transformVectors().

It additionally fixes behavior for edge cases.

## Testing

Tests included.

## Checklist

- [x] I, a human, have self-reviewed this PR and fully understand the changes within.
- [x] I have [made or updated tests](https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#test-guidelines) where possible.
- [x] All of my commits are properly scoped, compile successfully, and pass all tests.
- [x] This PR does not make sense to split up into smaller PRs.
- [x] An LLM was involved in the authoring of this code.

### CODE REVIEW

File: core/src/avm2/globals/flash/geom/matrix_3d.rs [hunk 1]
```diff
@@ -297,6 +297,108 @@ pub fn transpose<'gc>(
...
+
+    let mr = raw_data_of(this);
+
+    // 'vin' and 'vout' may be the same object, so borrows of their storage
+    // are scoped to a single access, never held across a read and a write.
+    let mut i = 0;
+    while i < result_vecs_length {
+        let x = vin.storage().get(i, activation)?.as_f64();
+        let y = vin.storage().get(i + 1, activation)?.as_f64();
+        let z = vin.storage().get(i + 2, activation)?.as_f64();
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/matrix_3d.rs]: It's probably better here to call `vin.storage()` just once per loop iteration and drop it once accesses are done.

File: core/src/avm2/globals/flash/geom/matrix_3d.rs [hunk 2]
```diff
@@ -297,6 +297,108 @@ pub fn transpose<'gc>(
...
+
+        let result = [
+            mr[0] * x + mr[4] * y + mr[8] * z + mr[12],
+            mr[1] * x + mr[5] * y + mr[9] * z + mr[13],
+            mr[2] * x + mr[6] * y + mr[10] * z + mr[14],
+        ];
+
+        for (offset, value) in result.into_iter().enumerate() {
+            vout.storage_mut(activation.gc())
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/matrix_3d.rs]: Similarly, this can call `vout.storage_mut()` once instead of thrice.

File: core/src/avm2/globals/flash/geom/matrix_3d.rs [hunk 3]
```diff
@@ -297,6 +297,108 @@ pub fn transpose<'gc>(
...
+    this: Value<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let this = this.as_object().unwrap();
+    let v = args.get_object(activation, 0, "vector")?;
+
+    let mr = raw_data_of(this);
+    let x = v.get_slot(vector3d_slots::X).as_f64();
+    let y = v.get_slot(vector3d_slots::Y).as_f64();
+    let z = v.get_slot(vector3d_slots::Z).as_f64();
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/matrix_3d.rs]: This should use `read_vector3d`

File: core/src/avm2/globals/flash/geom/matrix_3d.rs [hunk 4]
```diff
@@ -297,6 +297,108 @@ pub fn transpose<'gc>(
...
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/matrix_3d.rs]: This should use `read_vector3d`

Comment (kjarosh) [reply core/src/avm2/globals/flash/geom/matrix_3d.rs]: Done

File: core/src/avm2/globals/flash/geom/matrix_3d.rs (hunk 3 above)
Comment (kjarosh) [reply core/src/avm2/globals/flash/geom/matrix_3d.rs]: Done

File: core/src/avm2/globals/flash/geom/matrix_3d.rs (hunk 2 above)
Comment (kjarosh) [reply core/src/avm2/globals/flash/geom/matrix_3d.rs]: Done

File: core/src/avm2/globals/flash/geom/matrix_3d.rs (hunk 1 above)
Comment (kjarosh) [reply core/src/avm2/globals/flash/geom/matrix_3d.rs]: Done

**Review by Lord-McSweeney — APPROVED** [review]
